### PR TITLE
[Structural] NodalConcentratedElement - adding explicit dynamic functions

### PR DESCRIPTION
**Pull request by KlausBSautter**
+ Added the necessary functions for an explicit dynamic simulation to the NodalConcentratedElement
+ Added a test (basically repeating the implicit one and comparing results)

### CODE REVIEW

**Comment by KlausBSautter** [comment]
One question though:

in the code I see the following lines:

    const auto& rconst_this = *this;


    // Compute and add external forces
    const double nodal_mass = rconst_this.GetValue(NODAL_MASS);

I am doin this:

        double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
        r_nodal_mass += GetValue(NODAL_MASS);


It is working, but I dont understand why in the reference code the reference of the current element is obtained and then used to get the mass. Is this somehow better?

**Comment by loumalouomega** [comment]
> One question though:
> 
> in the code I see the following lines:
> 
> ```
> const auto& rconst_this = *this;
> 
> 
> // Compute and add external forces
> const double nodal_mass = rconst_this.GetValue(NODAL_MASS);
> ```
> 
> I am doin this:
> 
> ```
>     double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
>     r_nodal_mass += GetValue(NODAL_MASS);
> ```
> 
> It is working, but I dont understand why in the reference code the reference of the current element is obtained and then used to get the mass. Is this somehow better?

One is stored in the node, the other in the element. Try to follow the current implementation.

**Comment by loumalouomega** [comment]
CI is failing

**Comment by KlausBSautter** [comment]
> 
> 
> > One question though:
> > in the code I see the following lines:
> > ```
> > const auto& rconst_this = *this;
> > 
> > 
> > // Compute and add external forces
> > const double nodal_mass = rconst_this.GetValue(NODAL_MASS);
> > ```
> > 
> > 
> > I am doin this:
> > ```
> >     double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
> >     r_nodal_mass += GetValue(NODAL_MASS);
> > ```
> > 
> > 
> > It is working, but I dont understand why in the reference code the reference of the current element is obtained and then used to get the mass. Is this somehow better?
> 
> One is stored in the node, the other in the element. Try to follow the current implementation.

Yes that is obvious. Let me rephrase the question:
what is the difference between
`const auto& rconst_this = *this;`
`const double nodal_mass = rconst_this.GetValue(NODAL_MASS);`

and 
`const double nodal_mass = GetValue(NODAL_MASS);`

both access the element data and give the same result

**Comment by KlausBSautter** [comment]
> 
> 
> CI is failing

yes I forgot to get rid of matplotlib :D

**Comment by rubenzorrilla** [comment]
> > > One question though:
> > > in the code I see the following lines:
> > > ```
> > > const auto& rconst_this = *this;
> > > 
> > > 
> > > // Compute and add external forces
> > > const double nodal_mass = rconst_this.GetValue(NODAL_MASS);
> > > ```
> > > 
> > > 
> > > I am doin this:
> > > ```
> > >     double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
> > >     r_nodal_mass += GetValue(NODAL_MASS);
> > > ```
> > > 
> > > 
> > > It is working, but I dont understand why in the reference code the reference of the current element is obtained and then used to get the mass. Is this somehow better?
> > 
> > 
> > One is stored in the node, the other in the element. Try to follow the current implementation.
> 
> Yes that is obvious. Let me rephrase the question:
> what is the difference between
> `const auto& rconst_this = *this;`
> `const double nodal_mass = rconst_this.GetValue(NODAL_MASS);`
> 
> and
> `const double nodal_mass = GetValue(NODAL_MASS);`
> 
> both access the element data and give the same result

The first one is accessing the data from a const reference to the node, therefore no race conditions can occur if another element accesses the same node at the same time.

The second one is creating a const copy of the `NODAL_MASS` value.

**Comment by KlausBSautter** [comment]
@rubenzorrilla thanks. It accesses the element. So I better create a reference to the element itself and then call getvalue for that reference? So the first is safer ?
Then I will adjust the code .

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 1]
```diff
@@ -432,6 +433,63 @@ void NodalConcentratedElement::CalculateDampingMatrix(
...
+    const VectorType& rRHSVector,
+    const Variable<VectorType>& rRHSVariable,
+    Variable<double >& rDestinationVariable,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if (rDestinationVariable == NODAL_MASS) {
+        double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
+        r_nodal_mass += GetValue(NODAL_MASS);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]:
```suggestion
        #pragma omp atomic
        r_nodal_mass += GetValue(NODAL_MASS);
```

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 2]
```diff
@@ -432,6 +433,64 @@ void NodalConcentratedElement::CalculateDampingMatrix(
...
+            SizeType index = dimension * i;
+            array_1d<double, 3>& r_force_residual = GetGeometry()[i].FastGetSolutionStepValue(FORCE_RESIDUAL);
+            for (size_t j = 0; j < dimension; ++j) {
+                #pragma omp atomic
+                r_force_residual[j] += rRHSVector[index + j] - damping_residual_contribution[index + j];
+            }
+        }
+    } else if (rDestinationVariable == NODAL_INERTIA) {
+        double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
+        r_nodal_mass += GetValue(NODAL_MASS);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]:
```suggestion
        #pragma omp atomic
        r_nodal_mass += GetValue(NODAL_MASS);
```

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.hpp [hunk 3]
```diff
@@ -247,6 +247,17 @@ class NodalConcentratedElement
      */
     int Check(const ProcessInfo& rCurrentProcessInfo) override;
 
+    void AddExplicitContribution(
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.hpp:259]:
```suggestion
    /**
     * @brief This function is designed to make the element to assemble an rRHS vector identified by a variable rRHSVariable by assembling it to the nodes on the variable rDestinationVariable (double version)
     * @details The "AddEXplicit" FUNCTIONS THE ONLY FUNCTIONS IN WHICH AN ELEMENT IS ALLOWED TO WRITE ON ITS NODES.
     * The caller is expected to ensure thread safety hence SET/UNSETLOCK MUST BE PERFORMED IN THE STRATEGY BEFORE CALLING THIS FUNCTION
     * @param rRHSVector input variable containing the RHS vector to be assembled
     * @param rRHSVariable variable describing the type of the RHS vector to be assembled
     * @param rDestinationVariable variable in the database to which the rRHSVector will be assembled
     * @param rCurrentProcessInfo the current process info instance
     */
    void AddExplicitContribution(
```

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.hpp [hunk 4]
```diff
@@ -247,6 +247,26 @@ class NodalConcentratedElement
...
+     * @param rDestinationVariable variable in the database to which the rRHSVector will be assembled
+     * @param rCurrentProcessInfo the current process info instance
+     */
+    void AddExplicitContribution(
+        const VectorType& rRHSVector,
+        const Variable<VectorType>& rRHSVariable,
+        Variable<double >& rDestinationVariable,
+        const ProcessInfo& rCurrentProcessInfo) override;
+
+    void AddExplicitContribution(
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.hpp:274]:
```suggestion
    /**
     * @brief This function is designed to make the element to assemble an rRHS vector identified by a variable rRHSVariable by assembling it to the nodes on the variable (array_1d<double, 3>) version rDestinationVariable.
     * @details The "AddEXplicit" FUNCTIONS THE ONLY FUNCTIONS IN WHICH AN ELEMENT IS ALLOWED TO WRITE ON ITS NODES.
     * The caller is expected to ensure thread safety hence SET/UNSETLOCK MUST BE PERFORMED IN THE STRATEGY BEFORE CALLING THIS FUNCTION
     * @param rRHSVector input variable containing the RHS vector to be assembled
     * @param rRHSVariable variable describing the type of the RHS vector to be assembled
     * @param rDestinationVariable variable in the database to which the rRHSVector will be assembled
     * @param rCurrentProcessInfo the current process info instance
     */
    void AddExplicitContribution(
```

**Review by loumalouomega — DISMISSED** [review]
OK, I added some missing atomics and corrections

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla thanks. It accesses the element. So I better create a reference to the element itself and then call getvalue for that reference? So the first is safer ?
> Then I will adjust the code .

I thought you were accessing the nodes. Forget about my comment.

In this case I don't know if this has any advantage at all. I'm wondering if it is done in that way to avoid the parallel error in debug mode, that is to avoid doing the serial set in the initialize...

**Comment by philbucher** [comment]
I think @KlausBSautter akes abt the following:
difference btw
~~~c++
void SomeNONConstFct()
{
    // the following will ALLOCATE NODAL_MASS in the DataBase of the Element if it didn't exist before!
    double mass = this->GetValue(NODAL_MASS);

    // whereas this does not allocate bcs we have a const-reference:
    const auto& rconst_this = *this;
    double mass = rconst_this.GetValue(NODAL_MASS);

    // this would be an alternative (but is more cumbersome):
    double mass = 0.0;
    if (this->Has(NODAL_MASS) {
        mass = this->GetValue(NODAL_MASS);
    }
}

// and since we do not want to allocate, we first make a const-ref

**Review by philbucher — DISMISSED** [review]

**Comment by rubenzorrilla** [comment]
> I think @KlausBSautter akes abt the following:
> difference btw
> 
> ```c++
> void SomeNONConstFct()
> {
>     // the following will ALLOCATE NODAL_MASS in the DataBase of the Element if it didn't exist before!
>     double mass = this->GetValue(NODAL_MASS);
> 
>     // whereas this does not allocate bcs we have a const-reference:
>     const auto& rconst_this = *this;
>     double mass = rconst_this.GetValue(NODAL_MASS);
> 
>     // this would be an alternative (but is more cumbersome):
>     double mass = 0.0;
>     if (this->Has(NODAL_MASS) {
>         mass = this->GetValue(NODAL_MASS);
>     }
> }
> 
> // and since we do not want to allocate, we first make a const-ref
> ```

Now I see. Thanks for the clear explanation @philbucher .

File: applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp [hunk 5]
```diff
@@ -432,6 +433,65 @@ void NodalConcentratedElement::CalculateDampingMatrix(
...
+            array_1d<double, 3>& r_force_residual = GetGeometry()[i].FastGetSolutionStepValue(FORCE_RESIDUAL);
+            for (size_t j = 0; j < dimension; ++j) {
+                #pragma omp atomic
+                r_force_residual[j] += rRHSVector[index + j] - damping_residual_contribution[index + j];
+            }
+        }
+    } else if (rDestinationVariable == NODAL_INERTIA) {
+        double& r_nodal_mass = GetGeometry()[0].GetValue(NODAL_MASS);
+        #pragma omp atomic
+        r_nodal_mass += GetValue(NODAL_MASS);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/nodal_concentrated_element.cpp]: with C++17 you could do the following to be safe:
```suggestion
        r_nodal_mass += std::as_const(*this).GetValue(NODAL_MASS);
```

ref: https://en.cppreference.com/w/cpp/utility/as_const

**Review by philbucher — APPROVED** [review]
yes thats how it should be done
